Simplify `COALESCE` based on the nullability of its arguments

ranma42 opened the pull request:
Drop all of the arguments after the first non-nullable sub-expressions.
Simplify unary `COALESCE` to its argument.

roji commented on src/EFCore.SqlServer/Query/Internal/Translators/SqlServerStringAggregateMethodTranslator.cs:
```diff
@@ -105,7 +105,7 @@ public class SqlServerStringAggregateMethodTranslator : IAggregateMethodCallTran
                     enumerableArgumentIndex: 0,
                     nullable: true,
-                    argumentsPropagateNullability: new[] { false, true },
+                    argumentsPropagateNullability: new[] { false, false },
```
Thanks, good catch.

roji reviewed (commented):
This looks great - there's only the problem I mentioned in #33890:

> The main difficulty here is that we have various tests exercising Coalesce functionality, which are implemented over non-nullable columns; the Coalesce node would be stripped away there, and the tests would become useless. These need to be updated.

In other words this reduces our test coverage for coalesce but optimizing it away... Ideally we'd modify the tests to ensure that they actually exercise coalesce etc.

I'd also implement the SqlExpressionFactory (optimize at the source) as in #33890 at the same time (though it could be done separately too of course).

/cc @maumar

ranma42 commented:
> This looks great - there's only the problem I mentioned in #33890:
> 
> > The main difficulty here is that we have various tests exercising Coalesce functionality, which are implemented over non-nullable columns; the Coalesce node would be stripped away there, and the tests would become useless. These need to be updated.
> 
> In other words this reduces our test coverage for coalesce but optimizing it away... Ideally we'd modify the tests to ensure that they actually exercise coalesce etc.

I'll look into adding/updating tests before marking this as ready for review. Currently I am thinking of:
 - updating the tests that explicitly want a `COALESCE` so that it is not optimized away
 - adding some tests that check that n-ary `COALESCE` can be simplified to a `COALESCE` of a prefix of its arguments or just its first argument (with no `COALESCE` at all)

> 
> I'd also implement the SqlExpressionFactory (optimize at the source) as in #33890 at the same time (though it could be done separately too of course).

We can do that, but `SqlExpressionFactory` is currently lacking quite a lot of information to optimize as effectively as nullability can only be (trivially) computed for constants and columns, so optimizing in this location would still be useful.

This might change if `SqlExpression`s were extended with nullability information, as proposed in https://github.com/dotnet/efcore/issues/33889 (which would require more invasive changes).

roji commented:
> I'll look into adding/updating tests before marking this as ready for review. Currently I am thinking of:

Sounds great. Hopefully there are nullable alternatives in the model for the tests in question:::

> We can do that, but SqlExpressionFactory is currently lacking quite a lot of information to optimize as effectively as nullability can only be (trivially) computed for constants and columns, so optimizing in this location would still be useful.

Oh I completely agree - both optimizations make sense for sure. The SqlExpressionFactory really is more of a nice-to-have compared to the one implemented here in SqlNullabilityProcessor. I've updated #33890 to track both these things.

> This might change if SqlExpressions were extended with nullability information, as proposed in https://github.com/dotnet/efcore/issues/33889 (which would require more invasive changes).

Yeah, that's definitely a long-term discussion more than anything immediately actionable - we should proceed for now with the current query architecture regardless of that possible direction.

roji commented:
Assigning to myself but @maumar definitely let us know what you think etc.

maumar commented:
> > I'll look into adding/updating tests before marking this as ready for review. Currently I am thinking of:
> 
> Sounds great. Hopefully there are nullable alternatives in the model for the tests in question:::

I actually don't think we have a "natural" null value in a bool column in gears model. But you can always make one by doing a left join.

CogTag.Gear is optional 1:1 and there is a cog tag without corresponding gear, so 

`ss.Set<CogTag>().Where(x => (bool?)x.Gear.HasSoulPatch ?? false)` will exercise the code path and will be reflected in the results

ranma42 commented:
@roji I implemented the SqlExpressionFactory changes in https://github.com/dotnet/efcore/pull/34002 and based this PR on top of that (in the first PR I update the test so that they keep checking the appropriate COALESCE handling; in this PR I added tests for the optimization).

ranma42 commented:
The first commit belongs to #34078 (I rebased on top of that to have a working build). Apart from that, it is ready for review 🚀

ranma42 commented:
Rebased to resolve merge conflict

roji commented on src/EFCore.Relational/Query/SqlNullabilityProcessor.cs:
```diff
@@ -1422,12 +1422,33 @@ SqlExpression AddNullConcatenationProtection(SqlExpression argument, RelationalT
             nullable = coalesceNullable;
 
-            return sqlFunctionExpression.Update(sqlFunctionExpression.Instance, coalesceArguments);
+            if (coalesceArguments.Count == 1)
+            {
+                return coalesceArguments[0];
+            }
+            else if (coalesceArguments.Count == sqlFunctionExpression.Arguments.Count)
+            {
+                return sqlFunctionExpression.Update(sqlFunctionExpression.Instance, coalesceArguments);
+            }
+            else
+            {
+                return _sqlExpressionFactory.Function(
+                    sqlFunctionExpression.Name,
+                    coalesceArguments,
+                    sqlFunctionExpression.IsNullable,
+                    argumentsPropagateNullability: coalesceArguments.Select(_ => false).ToArray(),
+                    sqlFunctionExpression.Type,
+                    sqlFunctionExpression.TypeMapping);
+            }
```
nit: I'm a big fan of switches (especially expressions):

```suggestion
            return coalesceArguments switch
            {
                [var singleArgument] => singleArgument,

                _ when coalesceArguments.Count == sqlFunctionExpression.Arguments.Count
                    => sqlFunctionExpression.Update(sqlFunctionExpression.Instance, coalesceArguments),

                _ => _sqlExpressionFactory.Function(
                    sqlFunctionExpression.Name,
                    coalesceArguments,
                    sqlFunctionExpression.IsNullable,
                    argumentsPropagateNullability: coalesceArguments.Select(_ => false).ToArray(),
                    sqlFunctionExpression.Type,
                    sqlFunctionExpression.TypeMapping)
            };
```

roji commented on src/EFCore.Relational/Query/SqlNullabilityProcessor.cs:
```diff
@@ -1422,12 +1422,33 @@ SqlExpression AddNullConcatenationProtection(SqlExpression argument, RelationalT
+            }
+            else
+            {
+                return _sqlExpressionFactory.Function(
```
I think it makes sense to just have SqlFunctionExpression.Update accept an optional argumentsPropagateNullability argument

roji reviewed (approved):
Looks great as always, thanks. See two minor suggestions below.

roji commented on src/EFCore.Relational/Query/SqlExpressions/SqlFunctionExpression.cs:
```diff
@@ -331,9 +331,17 @@ public virtual SqlFunctionExpression ApplyTypeMapping(RelationalTypeMapping? typ
+        SqlExpression? instance,
+        IReadOnlyList<SqlExpression>? arguments,
+        IReadOnlyList<bool>? argumentsPropagateNullability = null)
+        => instance != Instance
```
We should do a pass to invert the logic of all of these across all expression types at some point (I hate negation + else)

ranma42 replied:
as we are handling this, should we also check that the arguments make sense?

Currently `foo.Update(foo.Instance, null)` always returns `foo` (which might not be intended).
Similarly, if `bar` is niladic (aka `bar.ArgumentsPropagateNullability` is `null`), `bar.Update(bar.Instance, [_factory.Constant(3)])` returns `bar` (`arguments` is ignored).

roji commented on src/EFCore.Relational/Query/SqlNullabilityProcessor.cs:
```diff
@@ -1422,12 +1422,25 @@ SqlExpression AddNullConcatenationProtection(SqlExpression argument, RelationalT
             nullable = coalesceNullable;
 
-            return sqlFunctionExpression.Update(sqlFunctionExpression.Instance, coalesceArguments);
+            return coalesceArguments switch
```
nit: now that this has only two arms, it can be a conditional operator

roji reviewed (commented):
One last tiny nit...

roji commented on src/EFCore.Relational/Query/SqlExpressions/SqlFunctionExpression.cs:
```diff
@@ -331,22 +331,49 @@ public virtual SqlFunctionExpression ApplyTypeMapping(RelationalTypeMapping? typ
+    {
+        argumentsPropagateNullability ??= ArgumentsPropagateNullability;
+
+        if (IsNiladic)
```
Shouldn't these checks be in the constructor?

ranma42 replied:
We could add them to the constructor if we want to, but they would not be sufficient to capture the misuse as mentioned in https://github.com/dotnet/efcore/pull/33938#discussion_r1656683191

Specifically, if either `arguments` or `Arguments` is `null`, the code would skip the constructor and just return `this`.

ranma42 replied:
@roji what is the best way forward? should I add checks both to the constructor and here?

roji replied:
Thanks for the ping.

I took another look... I'm not sure why SqlFunctionExpression has a separate IsNiladic property (I've been wanting to take a look at this for a while): the arguments being null (as opposed to empty) seems like the natural way to express a niladic function, and it indeed seems like whenever niladic is true, arguments is null and vice versa.

So unless I'm mistaken, we can probably just remove the IsNliadic property as an actual property (i.e. just make it return `Arguments is not null`). At that point, it's fine for Update() to switch to/from niladic - that's just an operation to change Arguments from null to non-null etc. And in any case argumentsPropagateNullability should correspond to arguments (both in nullness and in number of arguments when not null).

How does sounds?

ranma42 replied:
I especially like defining `IsNiladic` from arguments (and we can also annotate it so that the compiler knows about their relation).

I am not sure if it makes sense to ever switch between niladic (`FOOBAR`) and non-niladic (`FOOBAR()`, `FOOBAR(1, 2, 3)`, ...), but I guess it should not hurt (well, as long as the caller knows what it is doing 😅 ).

roji replied:
Yeah, I mean, if they can use Update() to switch between 1 to two arguments, or even to zero, I don't see why they shouldn't also switch to niladic... It's on them to get it right...

Is this something you want to do as part of this PR? Or separately? If you prefer I can submit a PR for the niladic change.

roji commented on src/EFCore.Relational/Query/SqlExpressions/SqlFunctionExpression.cs:
```diff
@@ -331,22 +309,35 @@ public virtual SqlFunctionExpression ApplyTypeMapping(RelationalTypeMapping? typ
+
+        return instance == Instance
+            && ((arguments == Arguments) || (arguments != null && Arguments != null && arguments.SequenceEqual(Arguments)))
+            && ((argumentsPropagateNullability == ArgumentsPropagateNullability)
```
OK, the problem this creates is that argumentsPropagateNullability being null could mean two things: either the user is switching to niladic (and so arguments is null too), or they want to just change the arguments without changing argumentsPropagateNullability (which seems like a common scenario).

Maybe we should just have two overloads, one only for changing arguments (doesn't have argumentsPropagateNullability parameter, and importantly, cannot change the number of arguments), and another with arguments and argumentsPropagateNullability?

ranma42 replied:
I'll do that (it was indeed my initial approach).
Note that the "common" scenario occurs just 3 times in the codebase (there are only 4 uses of this method)

roji commented on src/EFCore.Relational/Query/SqlExpressions/SqlFunctionExpression.cs:
```diff
@@ -240,7 +220,7 @@ public class SqlFunctionExpression : SqlExpression
     /// </summary>
     [MemberNotNullWhen(false, nameof(Arguments), nameof(ArgumentsPropagateNullability))]
-    public virtual bool IsNiladic { get; }
+    public virtual bool IsNiladic => Arguments is null;
```
Sorry, my bad - I misread.

roji reviewed (approved):
Thanks @ranma42!
